<a href="https://colab.research.google.com/github/rafafadilaras/Machine-Learning_2025/blob/main/Rafa_Fadil_Aras_JS13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Jobsheet 13 - Artificial Neural Network (ANN)**

## **Praktikum 1 - JST sederhana (2 layer) dengan forward pass dan backpropagation manual**

In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.26845216898128865
Epoch 1000, Loss: 0.23743952738078528
Epoch 2000, Loss: 0.18136189805229946
Epoch 3000, Loss: 0.14793584924684955
Epoch 4000, Loss: 0.13742788936747824
Epoch 5000, Loss: 0.1331412007220042
Epoch 6000, Loss: 0.13094205807660247
Epoch 7000, Loss: 0.12963513273465638
Epoch 8000, Loss: 0.1287792320890651
Epoch 9000, Loss: 0.12817938561855696
Prediksi:
[[0.0418    ]
 [0.49833313]
 [0.94833704]
 [0.50484248]]


### **Tugas 1**

1. Mengubah jumlah neuron hiden layer menjadi 3

In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.3068925545283221
Epoch 1000, Loss: 0.1940046599023882
Epoch 2000, Loss: 0.13501645305005108
Epoch 3000, Loss: 0.03944073939866964
Epoch 4000, Loss: 0.014921874020480876
Epoch 5000, Loss: 0.008219931925919046
Epoch 6000, Loss: 0.00545620174537408
Epoch 7000, Loss: 0.0040126499614300945
Epoch 8000, Loss: 0.00314365034475404
Epoch 9000, Loss: 0.002569605387275079
Prediksi:
[[0.02486432]
 [0.95593289]
 [0.95035328]
 [0.0603007 ]]


2. Membandingkan hasil loss dengan konfigurasi awal.

Berdasarkan hasil percobaan, dengan menambah jumlah neuron pada hidden layer dari 2 menjadi 3 mampu meningkatkan kemampuan belajar model. Pada konfigurasi awal (2 neuron), model gagal mempelajari pola XOR sepenuhnya dan terjebak dalam kondisi stuck, yang ditandai dengan nilai loss yang tertahan tinggi (0.128) dan prediksi angka yang ragu-ragu di sekitar 0.5. Sebaliknya, dengan menggunakan 3 neuron, jaringan memiliki kapasitas yang lebih besar dan fleksibel untuk memisahkan data. Hal ini membuat proses belajar menjadi lancar, menghasilkan loss yang turun drastis hingga mendekati nol (0.002), serta memberikan hasil prediksi yang sangat tegas dan akurat sesuai target.

3. Menambahkan fungsi aktivasi ReLU dan bandingkan hasil

In [1]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter (Tetap menggunakan 3 neuron hidden agar adil dengan percobaan sebelumnya)
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

np.random.seed(42) # Seed sama agar perbandingan fair
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# --- DEFINISI FUNGSI AKTIVASI BARU ---
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    # Turunan ReLU: 1 jika x > 0, selain itu 0
    return (x > 0).astype(float)

# Training
for epoch in range(10000):
    # --- FORWARD PASS ---
    z1 = np.dot(X, W1) + b1
    a1 = relu(z1)       # UBAH: Hidden layer sekarang pakai ReLU

    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)    # Output layer tetap Sigmoid

    # Hitung error
    error = y - a2

    # --- BACKPROPAGATION ---
    # Layer Output (Tetap Sigmoid)
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    # Layer Hidden (Sekarang ReLU)
    # Perhatikan: Turunan ReLU menggunakan z1 (input sebelum aktivasi)
    d_a1 = np.dot(d_a2, W2.T) * relu_derivative(z1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss:.6f}")

# Output akhir
print("\nPrediksi Akhir (ReLU di Hidden Layer):")
print(a2)

Epoch 0, Loss: 0.327478
Epoch 1000, Loss: 0.012797
Epoch 2000, Loss: 0.003283
Epoch 3000, Loss: 0.001731
Epoch 4000, Loss: 0.001147
Epoch 5000, Loss: 0.000846
Epoch 6000, Loss: 0.000666
Epoch 7000, Loss: 0.000547
Epoch 8000, Loss: 0.000462
Epoch 9000, Loss: 0.000400

Prediksi Akhir (ReLU di Hidden Layer):
[[0.02970009]
 [0.98605912]
 [0.98605934]
 [0.01163533]]


## **Praktikum 2**

In [9]:

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding
encoder = OneHotEncoder(sparse_output=False)
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Bangun model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(10, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(8, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

# Kompilasi
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model
model.fit(X_train, y_train, epochs=50, batch_size=8)

# Evaluasi
loss, acc = model.evaluate1(X_test, y_test)
print(f"Akurasi: {acc}")

Epoch 1/50


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


15/15 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.3069 - loss: 2.8660
Epoch 2/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.3069 - loss: 1.9525 
Epoch 3/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.2780 - loss: 1.4199 
Epoch 4/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.2860 - loss: 1.2523 
Epoch 5/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.3781 - loss: 1.2022 
Epoch 6/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3633 - loss: 1.1686 
Epoch 7/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.3226 - loss: 1.1485 
Epoch 8/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.2419 - loss: 1.1361 
Epoch 9/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.2499 - loss: 1.1265 
Epoch 10/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.2226 - loss: 1.1177 
Epoch 11/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.2226 - loss: 1.1089 
Epoch 12/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.2299 - los

### **Tugas 2**

Mengubah jumlah neuron hidden layer dan membandingkan akurasi dengan konfigurasi awal

In [16]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import tensorflow as tf

# Load dataset
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

# One-hot encoding
encoder = OneHotEncoder(sparse_output=False)
y = encoder.fit_transform(y)

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Bangun model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(12, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(10, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

# Kompilasi
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Latih model
model.fit(X_train, y_train, epochs=50, batch_size=8)

# Evaluasi
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi: {acc}")

Epoch 1/50


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


15/15 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.3448 - loss: 1.1097
Epoch 2/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3630 - loss: 1.0142 
Epoch 3/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5724 - loss: 0.9895 
Epoch 4/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5724 - loss: 0.9703 
Epoch 5/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6200 - loss: 0.9232 
Epoch 6/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8938 - loss: 0.8545 
Epoch 7/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8305 - loss: 0.8224 
Epoch 8/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8198 - loss: 0.7924 
Epoch 9/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8040 - loss: 0.7661 
Epoch 10/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8284 - loss: 0.7341 
Epoch 11/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8340 - loss: 0.7093 
Epoch 12/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8493 - los

Akurasi hasil konfigurasi awal (10, 8) = 9,337
Akurasi hasil setelah diubah (12, 10) = 1.0

### **Tugas 3**

Membandingkan Sigmoid vs ReLU pada dataset Iris

In [20]:
import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder

# Set seed agar hasil konsisten dan adil
tf.random.set_seed(42)
np.random.seed(42)

# --- 1. PREPARASI DATA ---
iris = load_iris()
X = iris.data
y = iris.target.reshape(-1, 1)

encoder = OneHotEncoder(sparse_output=False)
y = encoder.fit_transform(y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# --- 2. MEMBANGUN MODEL ---

# Fungsi pembantu untuk membuat model
def build_model(activation_type):
    model = tf.keras.Sequential([
        tf.keras.layers.Dense(10, activation=activation_type, input_shape=(4,)),
        tf.keras.layers.Dense(8, activation=activation_type),
        # Output layer tetap Softmax untuk klasifikasi multiclass
        tf.keras.layers.Dense(3, activation='softmax')
    ])
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

# Model A: Sigmoid
print("Training Model SIGMOID...")
model_sigmoid = build_model('sigmoid')
history_sigmoid = model_sigmoid.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)
loss_sig, acc_sig = model_sigmoid.evaluate(X_test, y_test, verbose=0)

# Model B: ReLU
print("Training Model RELU...")
model_relu = build_model('relu')
history_relu = model_relu.fit(X_train, y_train, epochs=50, batch_size=8, verbose=0)
loss_relu, acc_relu = model_relu.evaluate(X_test, y_test, verbose=0)

# --- 3. HASIL DAN VISUALISASI ---

print("\n" + "="*40)
print(f"{'MODEL':<15} | {'LOSS':<10} | {'AKURASI':<10}")
print("-" * 40)
print(f"{'Sigmoid':<15} | {loss_sig:.4f}     | {acc_sig:.4f}")
print(f"{'ReLU':<15} | {loss_relu:.4f}     | {acc_relu:.4f}")
print("="*40)

Training Model SIGMOID...


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Training Model RELU...

MODEL           | LOSS       | AKURASI   
----------------------------------------
Sigmoid         | 0.5654     | 1.0000
ReLU            | 0.3101     | 0.9000


Berdasarkan hasil percobaan pada dataset Iris, fungsi aktivasi ReLU terbukti lebih efisien dibandingkan Sigmoid. Model yang menggunakan ReLU menunjukkan penurunan loss yang jauh lebih tajam dan cepat pada awal proses training, yang menandakan bahwa model dapat belajar mengenali pola data dengan segera tanpa hambatan berarti. Sebaliknya, model dengan Sigmoid mengalami penurunan loss yang lebih lambat dan landai, hal ini disebabkan oleh karakteristik Sigmoid yang rentan terhadap masalah vanishing gradient (nilai gradien menjadi sangat kecil), sehingga bobot jaringan diperbarui secara perlahan